app/ide-desktop/lib/dashboard/src/authentication/src/http.tsx

```diff
@@ -0,0 +1,137 @@
+
+// ====================
+// === HttpResponse ===
+// ====================
+
+/** A wrapper around the built-in JavaScript HTTP {@link Response} type.
+ *
+ * This wrapper provides convenience methods like {@link HttpResponse.model} to deserialize the
+ * response body to a JSON object. */
+export class HttpResponse {
+  /** The wrapped HTTP {@link Response}. */
+  private response: Response;
+
+  /** Creates a new {@link HttpResponse} by wrapping the given {@link Response}. */
+  constructor(response: Response) {
+    this.response = response;
+  }
+
+  /** Returns the status code of the HTTP response. */
+  status = () => this.response.status;
+
+  /** Deserializes the HTTP response body as JSON, to the specified type. */
+  model = <T extends object>(): Promise<T> =>
+    this.response.json().then((json) => json as T);
+}
+
+// ======================
+// === RequestBuilder ===
+// ======================
+
+/** Builder for a new HTTP {@link RequestInitWithUrlAndHeaders}.
+ *
+ * Provides convenience methods for setting the request body, and sending the request using the HTTP
+ * {@link Client} provided when the builder was created. */
+class RequestBuilder {
+  /** HTTP client used to send the HTTP request, once it is created. */
+  private client: Client;
+  /** HTTP request being created. */
+  private request: RequestInitWithUrlAndHeaders;
+
+  /** Creates a new {@link RequestBuilder}. */
+  constructor(client: Client, url: string, config: RequestInit) {
+    const headers = new Headers();
+    this.client = client;
+    this.request = { ...config, url, headers };
+  }
+
+  /** Sets the provided JSON value as the body of the HTTP request, and adjusts the HTTP headers
+   * accordingly. */
+  json = (body: Json) => {
+    this.request.body = JSON.stringify(body);
+    this.request.headers.set("Content-Type", "application/json");
+    return this;
+  };
+
+  /** Finalizes the HTTP request and sends it using the HTTP client. */
+  send = () => this.client.executeRequest(this.request);
+}
+
+// === RequestInitWithUrlAndHeaders ===
+
+/** An HTTP request that is guaranteed to also have the URL it will be send to and an initialized
+ * map of {@link Headers}.
+ *
+ * This is an extension of the built-in {@link RequestInit} interface, which does not provide the
+ * aforementioned guarantees. In fact, it does not provide a `url` at all, and the `headers` field
+ * may even be `undefined` for that interface. */
+interface RequestInitWithUrlAndHeaders extends RequestInit {
+  url: string;
+  headers: Headers;
+}
+
+// ==============
+// === Client ===
+// ==============
+
+/** An HTTP client that can be used to create and send HTTP requests asynchronously. */
+export class Client {
+  /** A map of default headers that are included in every HTTP request sent by this client.
+   *
+   * This is useful for setting headers that are required for every request, like authentication
+   * tokens. */
+  defaultHeaders?: Headers;
```

Comment:
how/when is this assigned a value? you may want a constructor that takes `defaultParameters`